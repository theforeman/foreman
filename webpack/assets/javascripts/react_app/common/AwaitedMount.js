import PropTypes from 'prop-types';
import { useState, useEffect, useCallback } from 'react';
import store from '../redux';
import componentRegistry from '../components/componentRegistry';

// Mounts a component after all plugins have been imported to make sure that all plugins are available to the component
export const AwaitedMount = ({ component, data, flattenData }) => {
  const [mounted, setMounted] = useState(false);
  const [mountedComponent, setMountedComponent] = useState(null);
  const mountComponent = useCallback(async () => {
    if (componentRegistry.registry[component]) {
      setMounted(true);
      setMountedComponent(
        componentRegistry.markup(component, {
          data,
          store,
          flattenData,
        })
      );
    } else {
      // eslint-disable-next-line no-console
      console.debug(
        `Component not found: ${component}. The script for the component might not have been loaded yet.`
      );
    }
  }, [component, data, flattenData]);
  useEffect(() => {
    document.addEventListener('loadPlugin', mountComponent);
    return () => {
      window.removeEventListener('loadPlugin', mountComponent);
    };
  }, [mountComponent]);
  useEffect(() => {
    mountComponent();
  }, [mountComponent]);
  return mounted ? mountedComponent : null;
};

AwaitedMount.propTypes = {
  component: PropTypes.string.isRequired,
  data: PropTypes.object,
  flattenData: PropTypes.bool,
};
AwaitedMount.defaultProps = {
  data: {},
  flattenData: false,
};
