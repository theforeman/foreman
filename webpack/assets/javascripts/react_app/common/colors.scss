// colors
$pf-black-200: #ededed;
$pf-red-200: #a30000;
$pf-red-100: #c00;
$pf-orange-400: #ec7a08;
$pf-gold-400: #f0ab00;
$pf-green-400: #3f9c35;
$pf-green-600: #1e4f18;
$pf-blue-400: #0088ce;
$pf-color-white: #fff;
$pf-border-gray: #d1d1d1;

$switcher-max-height: 400px;
$switcher-max-width: 200px;
$switcher-min-width: 200px;

$nav-pf-vertical-active-icon-color: white;
$nav-pf-vertical-active-bg-color: #026991;
$nav-pf-vertical-secondary-active-bg-color: #024d6c;
$nav-pf-vertical-icon-color: #c7c7c7;
$nav-pf-vertical-secondary-indicator-color: #d1d1d1;
$nav-pf-vertical-bg-color: #024d6c;
