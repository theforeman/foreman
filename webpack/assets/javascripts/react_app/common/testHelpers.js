import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

export default {
  mockStorage: () => {
    const storage = {};

    return {
      setItem: (key, value) => {
        storage[key] = value || '';
      },
      getItem: key => (key in storage ? storage[key] : null),
      removeItem: key => {
        delete storage[key];
      },
      get length() {
        return Object.keys(storage).length;
      },
      key: i => {
        const keys = Object.keys(storage);

        return keys[i] || null;
      },
    };
  },
};

// a helper method for invoking a class method (for unit tests)
// obj = a class
// func = a tested function
// objThis = an object's this
// arg = function args

export const classFunctionUnitTest = (obj, func, objThis, args) =>
  obj.prototype[func].apply(objThis, args);

/**
 * Shallow render a component multipile times with fixtures
 * @param  {ReactComponent} Component Component to shallow-render
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 * @return {Object}                   key=fixture description, value=shallow-rendered component
 */
export const shallowRenderComponentWithFixtures = (Component, fixtures) =>
  Object.entries(fixtures).map(([description, props]) => ({
    description,
    component: shallow(<Component {...props} />),
  }));

/**
 * Test a component with fixtures and snapshots
 * @param  {ReactComponent} Component Component to test
 * @param  {Object}         fixtures  key=fixture description, value=props to apply
 */
export const testComponentSnapshotsWithFixtures = (Component, fixtures) =>
  shallowRenderComponentWithFixtures(Component, fixtures).forEach(
    ({ description, component }) =>
      it(description, () => expect(toJson(component)).toMatchSnapshot())
  );

/**
 * run an action (sync or async) and except the results to much snapshot
 * @param  {Function}  runAction  Action runner function
 * @return {Promise}
 */
export const testActionSnapshot = async runAction => {
  const actionResults = runAction();

  // if it's an async action
  if (typeof actionResults === 'function') {
    const dispatch = jest.fn();
    await actionResults(dispatch);

    expect(dispatch.mock.calls).toMatchSnapshot();
  } else {
    expect(actionResults).toMatchSnapshot();
  }
};

/**
 * Test actions with fixtures and snapshots
 * @param  {Object} fixtures key=fixture description, value=action runner function
 */
export const testActionSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, runAction]) =>
    it(description, () => testActionSnapshot(runAction))
  );

/**
 * Test a reducer with fixtures and snapshots
 * @param  {Function} reducer  reducer to test
 * @param  {Object}   fixtures key=fixture description, value=props to apply
 */
export const testReducerSnapshotWithFixtures = (reducer, fixtures) => {
  const reduce = ({ state, action = {} } = {}) => reducer(state, action);
  Object.entries(fixtures).forEach(([description, action]) =>
    it(description, () => expect(reduce(action)).toMatchSnapshot())
  );
};

/**
 * Test selectors with fixtures and snapshots
 * @param  {Object} fixtures  key=fixture description,
 *                            value=selector runner function
 */
export const testSelectorsSnapshotWithFixtures = fixtures =>
  Object.entries(fixtures).forEach(([description, selectorRunner]) =>
    it(description, () => expect(selectorRunner()).toMatchSnapshot())
  );
