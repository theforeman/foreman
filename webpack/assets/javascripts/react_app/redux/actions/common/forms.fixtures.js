export const requestData = {
  values: { a: 1 },
  url: '/api/resource',
  item: 'Resource',
};

export const requestDataMsg = Object.assign({}, requestData, {
  message: 'Customized success!',
});
