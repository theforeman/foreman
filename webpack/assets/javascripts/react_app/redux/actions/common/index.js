import API from '../../../API';

export const ajaxRequestAction = ({
  dispatch,
  requestAction,
  successAction,
  failedAction,
  url,
  item,
}) => {
  dispatch({ type: requestAction, payload: item });
  return API.get(url)
    .then(({ data }) =>
      dispatch({ type: successAction, payload: { ...item, ...data } })
    )
    .catch(error => dispatch({ type: failedAction, payload: { error, item } }));
};
