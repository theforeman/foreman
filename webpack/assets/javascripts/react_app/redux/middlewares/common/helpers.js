export const amendActionsPayload = (action, extPayload) => ({
  ...action,
  payload: {
    ...action.payload,
    ...extPayload,
  },
});
