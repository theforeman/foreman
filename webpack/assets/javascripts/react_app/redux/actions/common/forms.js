import { APIActions } from '../../API';
import { sprintf, translate as __ } from '../../../../react_app/common/I18n';

export class SubmissionError {
  constructor(errors) {
    this.errors = errors;
  }
}

const fieldErrors = ({ error }) => {
  const { errors, severity } = error;

  if (errors.base) {
    errors._error = {};
    errors._error.errorMsgs = errors.base;
    errors._error.severity = severity;
    delete errors.base;
  }

  return new SubmissionError(errors);
};

export const onError = error => {
  if (error.response.status === 422) {
    // Handle invalid form data
    throw fieldErrors(error.response.data);
  }
  throw new SubmissionError({
    _error: {
      errorMsgs: [
        `${__('Error submitting data:')} ${error.response.status} ${__(
          error.response.statusText
        )}`,
      ],
    },
  });
};

const verifyProps = (item, values) => {
  if (!item) {
    throw new Error('item must be defined, e.g. Bookmark');
  }
  if (!values) {
    throw new Error('values must be defined');
  }
};

export const submitForm = ({
  item,
  url,
  values: params,
  message,
  method = 'post',
  headers,
  apiActionTypes: actionTypes,
}) => {
  verifyProps(item, params);
  return async dispatch => {
    const uniqueAPIKey = `${item.toUpperCase()}_FORM_SUBMITTED`;

    const handleError = error => onError(error);

    const handleSuccess = ({ data }) => {
      dispatch({
        type: uniqueAPIKey,
        payload: { item, data },
      });
    };

    const successToast = response =>
      message || sprintf('%s was successfully created.', __(item));

    const errorToast = error =>
      sprintf(
        'Oh no! Something went wrong while submitting the form, the server returned the following error: %s',
        error
      );

    dispatch(
      APIActions[method]({
        key: uniqueAPIKey,
        url,
        headers,
        params,
        actionTypes,
        handleError,
        handleSuccess,
        successToast,
        errorToast,
      })
    );
  };
};
