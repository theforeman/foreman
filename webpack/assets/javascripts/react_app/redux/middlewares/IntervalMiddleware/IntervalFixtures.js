import { noop } from '../../../common/helpers';

export const key = 'SOME_KEY';
export const callback = noop;
export const args = [];
export const interval = 3000;
export const intervalID = 1212;
export const initialState = {};
export const stateWithKey = { [key]: intervalID };
export const fakeStore = {
  getState: () => ({
    intervals: initialState,
  }),
};
export const fakeStoreWithKey = {
  getState: () => ({
    intervals: stateWithKey,
  }),
};
