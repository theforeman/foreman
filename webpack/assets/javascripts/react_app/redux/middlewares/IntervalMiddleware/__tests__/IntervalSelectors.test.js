import { testSelectorsSnapshotWithFixtures } from '../../../../common/testHelpers';
import {
  selectIntervals,
  selectIntervalID,
  selectDoesIntervalExist,
} from '../IntervalSelectors';
import { key, stateWithKey } from '../IntervalFixtures';

const state = {
  intervals: stateWithKey,
};

const fixtures = {
  'should return the intervals wrapper': () => selectIntervals(state),
  'should return the interval ID': () => selectIntervalID(state, key),
  'should return if interval exists for a specific key': () =>
    selectDoesIntervalExist(state, key),
};

describe('intervals selectors', () =>
  testSelectorsSnapshotWithFixtures(fixtures));
