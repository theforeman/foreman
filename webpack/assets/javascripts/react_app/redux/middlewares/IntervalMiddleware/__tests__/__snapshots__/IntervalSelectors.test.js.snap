// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intervals selectors should return if interval exists for a specific key 1`] = `true`;

exports[`intervals selectors should return the interval ID 1`] = `1212`;

exports[`intervals selectors should return the intervals wrapper 1`] = `
Object {
  "SOME_KEY": 1212,
}
`;
